axum-macros/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+///     state: State,
+///     // and this via `TypedHeader::<ContentType>::from_request`
+///     #[from_request(via(TypedHeader))]
+///     content_type: ContentType,
+///     // Can still be combined with other extractors
+///     request_body: Bytes,
+/// }
+///
+/// #[derive(Clone)]
+/// struct State {
+///     // ...
+/// }
+///
+/// async fn handler(extractor: MyExtractor) {}
+/// ```
+///
+/// Note this requires the via extractor to be a generic tuple struct that implements `FromRequest`
+/// and has exactly one public field:
+///
+/// ```
+/// pub struct ViaExtractor<T>(pub T);
+///
+/// // impl<T, B> FromRequest<B> for ViaExtractor<T> { ... }
+/// ```
+///
+/// More complex via extractors are not supported and require writing a manual implementation.
+///
+/// ## Optional fields
+///
+/// `#[from_request(via(...))]` supports `Option<_>` and `Result<_, _>` to make fields optional:
+///
+/// ```
+/// use axum_macros::FromRequest;
+/// use axum::{
+///     extract::{TypedHeader, rejection::TypedHeaderRejection},
+///     headers::{ContentType, UserAgent},
+/// };
+///
+/// #[derive(FromRequest)]
+/// struct MyExtractor {
+///     // This will extracted via `Option::<TypedHeader<ContentType>>::from_request`
+///     #[from_request(via(TypedHeader))]
+///     content_type: Option<ContentType>,
+///     // This will extracted via
+///     // `Result::<TypedHeader<UserAgent>, TypedHeaderRejection>::from_request`
+///     #[from_request(via(TypedHeader))]
+///     user_agent: Result<UserAgent, TypedHeaderRejection>,
+/// }
+///
+/// async fn handler(extractor: MyExtractor) {}
+/// ```
+///
+/// ## The rejection
+///
+/// A rejection enum is also generated. It has a variant for each field:
+///
+/// ```
+/// use axum_macros::FromRequest;
+/// use axum::{
+///     extract::{Extension, TypedHeader},
+///     headers::ContentType,
+///     body::Bytes,
+/// };
+///
+/// #[derive(FromRequest)]
+/// struct MyExtractor {
+///     #[from_request(via(Extension))]
+///     state: State,
+///     #[from_request(via(TypedHeader))]
+///     content_type: ContentType,
+///     request_body: Bytes,
+/// }
+///
+/// // also generates
+/// //
+/// // #[derive(Debug)]
+/// // enum MyExtractorRejection {
+/// //     State(ExtensionRejection),
+/// //     ContentType(TypedHeaderRejection),
+/// //     RequestBody(BytesRejection),
+/// // }
+/// //
+/// // impl axum::response::IntoResponse for MyExtractor { ... }
+/// //
+/// // impl std::fmt::Display for MyExtractor { ... }
+/// //
+/// // impl std::error::Error for MyExtractor { ... }
+///
+/// #[derive(Clone)]
+/// struct State {
+///     // ...
+/// }
+/// ```
+///
+/// The rejection's `std::error::Error::source` implementation returns the inner rejection. This
+/// can be used to access source errors for example to customize rejection responses. Note this
+/// means the inner rejection types must themselves implement `std::error::Error`. All extractors
+/// in axum does this.
```

Comment:
I'm thinking if this will turn out to be annoying in practice. It means extractors that use `(StatusCode, String)` or `Response` as rejection will not work, since they don't implement `Error`. Perhaps the `Error`, `Display`, and `Debug` impls should be opt-in and disabled by default 🤔

Reply:
I think it should be on by default. You could have an opt-out of `#[from_request(rejection_derive(!Display, !Error))]` for opting out (and allow deriving extra traits for the rejection type too 😉 – [prior art](https://github.com/ruma/ruma/blob/37095f88553b311e7a70adaaabe39976fb8ff71c/crates/ruma-common/src/directory.rs#L135)).

Reply:
I like that!